Makefile

```diff
@@ -0,0 +1,93 @@
+#! /usr/bin/env make -f
+
+# This variable contains the first goal that matches any of the listed goals
+# here, else it contains an empty string. The net effect is to filter out
+# whether this current run of `make` requires a Python virtual environment.
+NEED_VENV := $(or \
+  $(findstring all,$(MAKECMDGOALS)), \
+  $(findstring quick-check,$(MAKECMDGOALS)), \
+  $(findstring check,$(MAKECMDGOALS)), \
+  $(findstring test,$(MAKECMDGOALS)), \
+  $(findstring dist,$(MAKECMDGOALS)), \
+  $(findstring bdist-wheel,$(MAKECMDGOALS)), \
+  $(findstring sdist,$(MAKECMDGOALS)), \
+  $(findstring docs,$(MAKECMDGOALS)) \
+)
+ifeq ($(NEED_VENV),)
+  # None of the current goals requires a virtual environment.
+else
+  ifeq ($(wildcard .venv),)
+    $(error No Python environment found, use `make setup` first)
+  else
+    PACKAGE_VERSION=$(shell .venv/bin/python -c 'import package; print(package.__version__)')
+  endif
+endif
+
+.PHONY: setup
+setup:
+	if [[ -d .venv ]]; then \
+	  echo "Python venv already exists, exiting" && exit 1; \
+	fi
+	if [[ -z "${PYTHON}" ]]; then \
+	  python3.10 -m venv --upgrade-deps .venv; \
+	else \
+	  ${PYTHON} -m venv --upgrade-deps .venv; \
+	fi
+	. .venv/bin/activate && \
+	pip install --upgrade pip && \
+	pip install --editable .[hooks,dev,test,docs]
+	. .venv/bin/activate && \
+	pre-commit install && \
+	pre-commit install --hook-type commit-msg && \
+	pre-commit install --hook-type pre-push
+
+.PHONY: all
+all: check test dist docs
+
+.PHONY: quick-check check
+quick-check:
+	. .venv/bin/activate && \
+	pre-commit run pylint --all-files && \
+	pre-commit run mypy --all-files
+check:
+	. .venv/bin/activate && \
+	pre-commit run --all-files
+
+.PHONY: test
+test:
+	. .venv/bin/activate && \
+	pre-commit run pytest --hook-stage push
```

Comment:
We _should_ run
```
python -W always::DeprecationWarning -m pytest
```
to make sure warnings are issued, see also the [`-W` command-line option](https://docs.python.org/3/using/cmdline.html#cmdoption-W).